openfisca_core/formulas.py

```diff
@@ -954,48 +968,84 @@ def set_input_dispatch_by_period(formula, period, array):
                 sub_period = sub_period.offset(1)
-            if sub_periods_count > 0:
-                divided_array = remaining_array / sub_periods_count
-                sub_period = period.start.period(period_unit)
-                while sub_period.start < after_instant:
-                    if holder.get_array(sub_period) is None:
-                        holder.put_in_cache(divided_array, sub_period)
-                    sub_period = sub_period.offset(1)
-        if period_unit == u'year':
-            remaining_array = array.copy()
-            month = period.start.period(u'month')
-            months_count = 12 * period_size
-            while month.start < after_instant:
-                existing_array = holder.get_array(month)
-                if existing_array is not None:
-                    remaining_array -= existing_array
-                    months_count -= 1
-                month = month.offset(1)
-            if months_count > 0:
-                divided_array = remaining_array / months_count
-                month = period.start.period(u'month')
-                while month.start < after_instant:
-                    if holder.get_array(month) is None:
-                        holder.put_in_cache(divided_array, month)
-                    month = month.offset(1)
+        else:
+            raise ValueError('Values for variable {} are already set on period {}.'.format(holder.column.name, period))
+
+    elif behavior == SET_INPUT_REPLACE:
+        # Count the number of elementary periods to change
+        sub_period = period.start.period(cached_period_unit)
+        sub_periods_count = 0
+        while sub_period.start < after_instant:
+            sub_periods_count += 1
+            sub_period = sub_period.offset(1)
+
+        # Cache the input data
+        divided_array = array / sub_periods_count
+        sub_period = period.start.period(cached_period_unit)
+        while sub_period.start < after_instant:
+            holder.put_in_cache(divided_array, sub_period)
+            sub_period = sub_period.offset(1)
+
+    elif behavior == SET_INPUT_ADD:
+        # Count the number of elementary periods to change
+        sub_period = period.start.period(cached_period_unit)
+        sub_periods_count = 0
+        while sub_period.start < after_instant:
+            sub_periods_count += 1
+            sub_period = sub_period.offset(1)
+
+        # Cache the input data
+        divided_array = array / sub_periods_count
+        sub_period = period.start.period(cached_period_unit)
+        while sub_period.start < after_instant:
+            old_array = holder.get_array(sub_period)
+            if old_array is None:
+                holder.put_in_cache(divided_array, sub_period)
+            else:
+                holder.put_in_cache(divided_array + old_array, sub_period)
+            sub_period = sub_period.offset(1)
+
+    else:
+        raise ValueError('Unknown behavior {}'.format(behavior))
```

Comment:
Even if I use an enum, as python is not strongly typed, I can still give a bad argument to the function no ?